master/internal/logretention/logretention.go

```diff
@@ -0,0 +1,115 @@
+package logretention
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-co-op/gocron/v2"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+const retainForever = -1
+
+var (
+	log = logrus.WithField("component", "log-retention")
+
+	schedulerDefaultOpts = []gocron.SchedulerOption{
+		gocron.WithLimitConcurrentJobs(1, gocron.LimitModeReschedule),
+	}
+	scheduler gocron.Scheduler
+
+	// WaitGroup is used for testing purposes to wait for the log retention scheduler to finish.
+	WaitGroup *sync.WaitGroup
+)
+
+func init() {
+	SetupScheduler()
+}
+
+// SetupScheduler creates a new scheduler with the provided options.
+// Should only be called by init() or test functions, and will panic on error.
+func SetupScheduler(opts ...gocron.SchedulerOption) {
+	schedulerOpts := append([]gocron.SchedulerOption{}, schedulerDefaultOpts...)
+	schedulerOpts = append(schedulerOpts, opts...)
+
+	newScheduler, err := gocron.NewScheduler(schedulerOpts...)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to create logretention scheduler"))
+	}
+	scheduler = newScheduler
+}
+
+// Schedule begins a log deletion schedule according to the provided LogRetentionPolicy.
+func Schedule(config model.LogRetentionPolicy) error {
+	// Create a task that deletes expired task logs.
+	task := gocron.NewTask(func() {
+		defer func() {
+			if WaitGroup != nil {
+				WaitGroup.Done()
+			}
+		}()
+		count, err := DeleteExpiredTaskLogs(config.Days)
+		if err != nil {
+			log.WithError(err).Error("failed to delete expired task logs")
+		} else if count > 0 {
+			log.WithField("count", count).Info("deleted expired task logs")
+		}
+	})
+	// If a cleanup schedule is set, schedule the cleanup task.
+	if config.Schedule != nil {
+		if d, err := time.ParseDuration(*config.Schedule); err == nil {
+			// Try to parse out a duration.
+			log.WithField("duration", d).Debug("running task log cleanup with duration")
+			_, err := scheduler.NewJob(gocron.DurationJob(d), task)
+			if err != nil {
+				return errors.Wrapf(err, "failed to schedule duration task log cleanup")
+			}
+		} else {
+			// Otherwise, use a cron.
+			log.WithField("cron", *config.Schedule).Debug("running task log cleanup with cron")
+			_, err := scheduler.NewJob(gocron.CronJob(*config.Schedule, false), task)
+			if err != nil {
+				return errors.Wrapf(err, "failed to schedule cron task log cleanup")
+			}
+		}
+	}
+	// Start the scheduler.
+	scheduler.Start()
+	return nil
+}
+
+// DeleteExpiredTaskLogs deletes task logs older than days time when defined and non-negative.
+// Task configured values may override the default provided number of days for retention.
+func DeleteExpiredTaskLogs(days *int16) (int64, error) {
```

Comment:
take a context?

Reply:
I will make this change but I think we will just end up passing context.Background in. In the sync api case it might make sense to pass in the call context but if it is async then would the background context anyways.